### PR TITLE
Add cost estimate to LLM scan

### PR DESCRIPTION
**Pull request by mattbit**

### CODE REVIEW

**Comment by rabah-khalek** [comment]
is this still WIP?

**Comment by mattbit** [comment]
@rabah-khalek added measured usage of LLM API

File: giskard/scanner/scanner.py [hunk 1]
```diff
@@ -183,6 +148,64 @@ def _postprocess(self, issues: Sequence[Issue]) -> Sequence[Issue]:
...
+            if not model.meta.name or not model.meta.description:
+                raise ValueError(
+                    "LLM models must have a name and a description. Please make sure the `name` and `description` "
+                    "parameters of your model are not empty. We will use them to configure our detectors."
+                )
+            if not model.meta.feature_names:
+                raise ValueError(
+                    "LLM models must specify their input variables to be analyzed. "
+                    "Please make sure to set the `feature_names` parameter when wrapping your model."
+                )
```
Comment (rabah-khalek) [inline giskard/scanner/scanner.py:175]: let's remove these in favor of https://github.com/Giskard-AI/giskard/pull/1525

File: giskard/scanner/scanner.py [hunk 2]
```diff
@@ -183,6 +148,64 @@ def _postprocess(self, issues: Sequence[Issue]) -> Sequence[Issue]:
...
+            raise ValueError(
+                "The model object you provided is not valid. Please wrap it with the `giskard.Model` class. "
+                "See the instructions here: https://docs.giskard.ai/en/latest/guides/wrap_model/index.html"
+            )
+
+        if dataset is not None and not isinstance(dataset, Dataset):
+            raise ValueError(
+                "The dataset object you provided is not valid. Please wrap your dataframe with `giskard.Dataset`. "
+                "You can follow the docs here: https://docs.giskard.ai/en/latest/guides/wrap_dataset/index.html"
+            )
```
Comment (rabah-khalek) [inline giskard/scanner/scanner.py:162]: let's move this to `model_validation` and `dataset_validation`

File: giskard/scanner/scanner.py [hunk 3]
```diff
@@ -183,6 +148,64 @@ def _postprocess(self, issues: Sequence[Issue]) -> Sequence[Issue]:
...
+            warning(
+                "You dataframe has duplicate indexes, which is currently not supported. "
+                "We have to reset the dataframe index to avoid issues."
+            )
+            dataset = Dataset(
+                df=dataset.df.reset_index(drop=True),
+                name=dataset.name,
+                target=dataset.target,
+                column_types=dataset.column_types,
+            )
```
Comment (rabah-khalek) [inline giskard/scanner/scanner.py:196]: I think that is generally problematic for us, so i would move it to `dataset_validation`

**Review by rabah-khalek — CHANGES_REQUESTED** [review]
it looks good, can we just factor out the validation to where it belongs?

I tried it on couple of detectors (for the reader’s comment notebook):
- Harmfulness 
```
Estimated calls to your model: ~20
Estimated OpenAI GPT-4 calls for evaluation: 25 (~9461.0 prompt tokens and ~800.0 sampled tokens)
OpenAI API costs for evaluation are estimated to $0.33.

OpenAI GPT-4 calls for evaluation: 25 (10146 prompt tokens and 773 sampled tokens)
OpenAI API costs for evaluation amount to $0.35 (standard pricing).
```

- misinformation
```
Estimated calls to your model: ~30
Estimated OpenAI GPT-4 calls for evaluation: 22 (~9148.0 prompt tokens and ~1110.0 sampled tokens)
OpenAI API costs for evaluation are estimated to $0.34.

OpenAI GPT-4 calls for evaluation: 22 (9003 prompt tokens and 1072 sampled tokens)
OpenAI API costs for evaluation amount to $0.33 (standard pricing).
```

P.S. With the understanding that the cost estimate is rough and tailored today to what we have as evaluation prompts, it looks okay to me. It could easily get off charts if we change drastically the evaluation in the llm-as-a-judge.

File: giskard/scanner/scanner.py (hunk 1 above)
Comment (mattbit) [reply giskard/scanner/scanner.py:175]: out of scope for this PR, validation should be handled in #1525

File: giskard/scanner/scanner.py (hunk 2 above)
Comment (mattbit) [reply giskard/scanner/scanner.py:162]: This belongs to the scan, it checks that the arguments coming in are instances of the right objects.

File: giskard/scanner/scanner.py (hunk 3 above)
Comment (mattbit) [reply giskard/scanner/scanner.py:196]: out of scope for this PR

File: giskard/scanner/llm/llm_prompt_injection_detector.py [hunk 4]
```diff
@@ -30,9 +30,13 @@ class LLMPromptInjectionDetector(Detector):
     .. [#] Leon Derczynsky, garak:  LLM vulnerability scanner, https://github.com/leondz/garak
     """
 
-    def __init__(self, threshold: float = 0.5, num_samples=100):
+    def __init__(self, threshold: float = 0.5):
         self.threshold = threshold  # default
-        self.num_samples = num_samples
```
Comment (mattbit) [inline giskard/scanner/llm/llm_prompt_injection_detector.py:35]: @rabah-khalek could you double check that this makes sense?

Comment (rabah-khalek) [reply giskard/scanner/llm/llm_prompt_injection_detector.py:35]: Yes, `num_samples` is only actively used in https://github.com/Giskard-AI/giskard/pull/1533/. It's a nice feature to have, but not a blocker for the release.

File: giskard/scanner/scanner.py (hunk 2 above)
Comment (rabah-khalek) [reply giskard/scanner/scanner.py:193]: yeah well that's what the validation step does

**Review by rabah-khalek — APPROVED** [review]

**Pull request merged**